include/vinecopulib/vinecop/implementation/class.ipp

```diff
@@ -1087,59 +1082,110 @@ Vinecop::get_npars() const
   size_t trunc_lvl = rvine_structure_.get_trunc_lvl();
   auto order = rvine_structure_.get_order();
   auto inverse_order = tools_stl::invert_permutation(order);
+  auto disc_cols = tools_select::get_disc_cols(var_types_);
 
   // fill first row of hfunc2 matrix with evaluation points;
   // points have to be reordered to correspond to natural order
-  Eigen::MatrixXd hfunc1(n, d);
-  Eigen::MatrixXd hfunc2(n, d);
-  for (size_t j = 0; j < d; ++j)
+  Eigen::MatrixXd hfunc1(n, d), hfunc2(n, d), hfunc1_sub(n, d), hfunc2_sub(n, d);
+  for (size_t j = 0; j < d; ++j) {
     hfunc2.col(j) = u.col(order[j] - 1);
+  }
+  if (is_discrete()) {
+    hfunc1_sub = hfunc1;
+    hfunc2_sub = hfunc2;  
+  }
+
+  // fill first row of hfunc2 matrix with evaluation points;
+  // points have to be reordered to correspond to natural order
+  for (size_t j = 0; j < d_; ++j) {
+    hfunc2.col(j) = u.col(order[j] - 1);
+    if (var_types_[order[j] - 1] == "d") {
+      hfunc2_sub.col(j) = u.col(d_ + disc_cols[order[j] - 1]);
+    }
+  }
 
   auto do_batch = [&](const tools_batch::Batch& b) {
-    Eigen::MatrixXd u_e(b.size, 2);
+    Eigen::MatrixXd u_e, u_e_sub;
     for (size_t tree = 0; tree < trunc_lvl; ++tree) {
       tools_interface::check_user_interrupt(
         static_cast<double>(n) * static_cast<double>(d) > 1e5);
       for (size_t edge = 0; edge < d - tree - 1; ++edge) {
         tools_interface::check_user_interrupt(edge % 100 == 0);
         // extract evaluation point from hfunction matrices (have been
         // computed in previous tree level)
+        Bicop* edge_copula = &pair_copulas_[tree][edge];
+        auto var_types = edge_copula->get_var_types();
         size_t m = rvine_structure_.min_array(tree, edge);
+
+        u_e = Eigen::MatrixXd(b.size, 2);
         u_e.col(0) = hfunc2.block(b.begin, edge, b.size, 1);
         if (m == rvine_structure_.struct_array(tree, edge, true)) {
           u_e.col(1) = hfunc2.block(b.begin, m - 1, b.size, 1);
         } else {
           u_e.col(1) = hfunc1.block(b.begin, m - 1, b.size, 1);
         }
 
+        if ((var_types[0] == "d") || (var_types[1] == "d")) {
+          u_e.conservativeResize(b.size, 4);
+          u_e.col(2) = hfunc2_sub.block(b.begin, edge, b.size, 1);
+          if (m == rvine_structure_.struct_array(tree, edge, true)) {
+            u_e.col(3) = hfunc2_sub.block(b.begin, m - 1, b.size, 1);
+          } else {
+            u_e.col(3) = hfunc1_sub.block(b.begin, m - 1, b.size, 1);
+          }
+        }
+
         // h-functions are only evaluated if needed in next step
-        Bicop edge_copula = pair_copulas_[tree][edge].as_continuous();
         if (rvine_structure_.needed_hfunc1(tree, edge)) {
-          hfunc1.block(b.begin, edge, b.size, 1) = edge_copula.hfunc1(u_e);
+          hfunc1.block(b.begin, edge, b.size, 1) = edge_copula->hfunc1(u_e);
+          if (var_types[1] == "d") {
+            u_e_sub = u_e;
+            u_e_sub.col(1) = u_e.col(3);
```

Comment:
Are we sure that the line above is doing a deep copy and thus that this line is not also modifying u_e? Otherwise, stuff at line 1183 becomes incorrect.

Reply:
yes, `u_e_sub` and `u_e` are two Eigen::MatrixXd objects with separate storage. There would be no deep copy when declarying `auto u_e_sub = u_e;`, but we're not doing that.